mitosheet/src/mito/components/endo/ColumnHeaders.tsx

```diff
@@ -88,6 +89,11 @@ const ColumnHeaders = (props: {
                         e.preventDefault()
                         e.persist()
 
+                        // First we handle highlighting the border
+                        const {columnIndex} = getIndexesFromXAndY(e.clientX, e.clientY)
+                        props.setUIState({ ...props.uiState, highlightedColumnIndex: columnIndex });
```

Comment:
Every time we update the UIState, the entire grid gets rerendered, so this is working well on small datasets, but when I tested it on a larger dataset, it becomes really laggy. 

Here's a few refactor ideas that might make this more performant: 

1. One simple idea is to only update the `highlighitedColumnIndex` if it changes. That should reduce a ton of the rerenders! 
2. Instead of using the UIState to keep track of the highlightedColumnIndex, I wonder if we could instead just keep track that we're in `reorder mode` and apply a class like `highlight-border` to the column that is hovered over. The challenge here would be figuring out based on where the cursor is what column to apply the class too because we don't have access to the drag event in `GridData.tsx` and it would be hard to link the column header that we are hovering over to the column, I think.. 

There's a few large datasets in the monorepo that you can use for testing purposes -- you can find them in the mitosheet/datasets/ folder

Reply:
Haven't looked at the code (so excuse me popping in here), but another idea here I thought I'd throw out -- would it be possible to just use state that we already have to get the reorder mode information?

If we already store the dragged column index somewhere, we could just highlight whatever it is over potentially (I am not sure we do -- so this might just be silly). 

Another possible optimization here might be storing the highlightedColumnIndex within the ColumnHeaders component -- so only this rerenders -- rather than in the UI state.

Reply:
> Every time we update the UIState, the entire grid gets rerendered, so this is working well on small datasets, but when I tested it on a larger dataset, it becomes really laggy.

I'm having some trouble reproducing the laggy-ness (even in the `datasets/large-datasets` directory), is there a specific spreadsheet you're seeing that behavior for? Maybe it's the way I'm doing it? 

> Here's a few refactor ideas that might make this more performant:
> 
> 1. One simple idea is to only update the `highlighitedColumnIndex` if it changes. That should reduce a ton of the rerenders!

Just tried that! It worked well for me so I'll try pushing that in case it fixes the laggy-ness for you

> 2. Instead of using the UIState to keep track of the highlightedColumnIndex, I wonder if we could instead just keep track that we're in `reorder mode` and apply a class like `highlight-border` to the column that is hovered over. The challenge here would be figuring out based on where the cursor is what column to apply the class too because we don't have access to the drag event in `GridData.tsx` and it would be hard to link the column header that we are hovering over to the column, I think..

Maybe I'm misunderstanding, but I think the issue with this is that the cells are rendered as rows, so we don't have a "column" to add a drag handler / class onto. So I think we'd need to re-render either way? But I'm not totally sure! 

Also - we do keep track of a "reorder mode" but I'm not sure if it helps much because it mostly keeps track of whether we're readjusting the size of the column or reordering the columns. 

> would it be possible to just use state that we already have to get the reorder mode information?

Yep! See above answer 
 
> If we already store the dragged column index somewhere, we could just highlight whatever it is over potentially (I am not sure we do -- so this might just be silly).

We're not keeping track of it that I know of - we're using the `onDrop` function to handle the actual re-ordering, so what happens while you're dragging isn't really involved with that.  

> Another possible optimization here might be storing the highlightedColumnIndex within the ColumnHeaders component -- so only this rerenders -- rather than in the UI state.

We need that variable for all of the cells in that column to also have a highlighted border - unless we want to only highlight the border between column headers? I could see an argument for that

Reply:
Actually just checked and it only works when you drop the header on the same level as a header, so I think it might be more intuitive if we only highlight the border between headers - should I go ahead and add that?

Reply:
Ok just tried that out and it's a lot harder to see - especially when you're trying to drop the column on top of the header already. 
<img width="1440" alt="image" src="https://github.com/mito-ds/monorepo/assets/6673460/424ed3fb-3571-4bf0-958e-dbcd41c8daa2">

Reply:
Update from conversation with Aaron: the lagginess is fixed but we noticed that when you drop the header on the *grid*, the highlighted column ends up staying highlighted. So I'm going to try to add an `onDragOver` or `onDrop` handler for the `EndoGrid` and see if I can fix it up. In another PR, we might try to update the logic to allow users to drop the column header over the `EndoGrid` to reorder columns.